lib/mayaUsd/ufe/UsdContextOps.cpp

```diff
@@ -173,6 +192,137 @@ struct WaitCursor
+        const std::vector<std::string> vettedSurfaces = { "ND_standard_surface_surfaceshader",
+                                                          "ND_gltf_pbr_surfaceshader",
+                                                          "ND_UsdPreviewSurface_surfaceshader" };
+        for (auto&& identifier : vettedSurfaces) {
+            auto shaderDef = sdrRegistry.GetShaderNodeByIdentifier(TfToken(identifier));
+            if (!shaderDef) {
+                continue;
+            }
+            std::string label = UsdMayaUtil::prettifyName(shaderDef->GetFamily().GetString());
+            mxSurfaceShaders.emplace_back(label, shaderDef->GetIdentifier().GetString());
+        }
+    }
+    return mxSurfaceShaders;
+}
+
+//! \brief Create an Item and select it:
+class BaseCreateAndSelectUndoableCommand : public Ufe::UndoableCommand
+{
+public:
+    BaseCreateAndSelectUndoableCommand(const Ufe::UndoableCommand::Ptr& creationCmd)
+        : _creationCmd(creationCmd)
+    {
+    }
+
+    void redo() override
+    {
+        if (_selectionCmd) {
+            _creationCmd->redo();
+            _selectionCmd->redo();
+        }
+    }
+
+    void undo() override
+    {
+        if (_selectionCmd) {
+            _selectionCmd->undo();
+            _creationCmd->undo();
+        }
+    }
+
+protected:
+    Ufe::UndoableCommand::Ptr _creationCmd;
+    Ufe::UndoableCommand::Ptr _selectionCmd;
+};
+
+//! \brief Create a Prim and select it:
+class UsdUndoAddNewPrimAndSelectCommand : public BaseCreateAndSelectUndoableCommand
+{
+public:
+    UsdUndoAddNewPrimAndSelectCommand(
+        const MAYAUSD_NS::ufe::UsdUndoAddNewPrimCommand::Ptr& creationCmd)
+        : BaseCreateAndSelectUndoableCommand(creationCmd)
+    {
+    }
+
+    void execute() override
+    {
+        _creationCmd->execute();
+        auto addPrimCmd
+            = std::dynamic_pointer_cast<MAYAUSD_NS::ufe::UsdUndoAddNewPrimCommand>(_creationCmd);
+        // Create the selection command only if the creation succeeded:
+        if (!addPrimCmd->newUfePath().empty()) {
+            Ufe::Selection newSelection;
+            newSelection.append(Ufe::Hierarchy::createItem(addPrimCmd->newUfePath()));
+            _selectionCmd = Ufe::SelectionReplaceWith::createAndExecute(
+                Ufe::GlobalSelection::get(), newSelection);
+        }
+    }
+};
+
+//! \brief Create a working Material and select it:
+class InsertChildAndSelectCommand : public BaseCreateAndSelectUndoableCommand
+{
+public:
+    InsertChildAndSelectCommand(const Ufe::InsertChildCommand::Ptr& creationCmd)
+        : BaseCreateAndSelectUndoableCommand(creationCmd)
+    {
+    }
+
+    void execute() override
+    {
+        _creationCmd->execute();
+        auto insertChildCmd = std::dynamic_pointer_cast<Ufe::InsertChildCommand>(_creationCmd);
+        // Create the selection command only if the creation succeeded:
+        if (insertChildCmd->insertedChild()) {
```

Comment:
We expect that the creation command will report an error if it fails. Should composite commands report errors on top or other errors?

The question is good though, because this is exactly what I am doing in the composite UsdUndoAssignNewMaterialCommand. I should probably go there and remove all TF_RUNTIME_ERROR since each individual command would be expected to properly report, and it then becomes unnecessary to complain that the error in AddNewPrim is causing AssignNewMaterial to fail.